docs/csharp/methods.md

```diff
@@ -1,27 +1,295 @@
+
+> [!NOTE]
+> An async method returns to the caller when either it encounters the first awaited object that’s not yet complete or it gets to the end of the async method, whichever occurs first.
+
+An async method can have a return type of @System.Threading.Tasks.Task<TResult>, @System.Threading.Tasks.Task, or `void`. The `void` return type is used primarily to define event handlers, where a `void` return type is required. An async method that returns `void` can't be awaited, and the caller of a void-returning method can't catch exceptions that the method throws.
+
+In the following example, `DelayAsync` is an async method that has a return statement that returns an integer. Because it is an async method, its method declaration must have a return type of `Task<int>`. Because the return type is `Task<int>`, the evaluation of the `await` expression in `DoSomethingAsync` produces an integer, as the following `int result = await delayTask` statement demonstrates.
+
+[!CODE [csSnippets.Methods#102](../../samples/snippets/csharp/concepts/methods/async1.cs#102)]
+
+An async method can't declare any [ref](https://msdn.microsoft.com/en-us/library/14akc2c7.aspx) or [out](https://msdn.microsoft.com/en-us/library/t3c3bfhx.aspx) parameters, but it can call methods that have such parameters.
+
+ For more information about async methods, see [Asynchronous Programming with Async and Await](https://msdn.microsoft.com/en-us/library/mt674882.aspx), [Control Flow in Async Programs](https://msdn.microsoft.com/en-us/library/mt674892.aspx), and [Async Return Types](https://msdn.microsoft.com/en-us/library/mt674893.aspx).
+
+## <a name="expr" /> Expression body definitions ##
+
+It is common to have method definitions that simply return immediately with the result of an expression, or that have a single statement as the body of the method.  There is a syntax shortcut for defining such methods using `=>`:
+
+```c#
+
+public Point Move(int dx, int dy) => new Point(x + dx, y + dy);
+public void Print() => Console.WriteLine(First + " " + Last);
+// Works with operators, properties, and indexers too.
+public static Complex operator +(Complex a, Complex b) => a.Add(b);
+public string Name => First + " " + Last;
+public Customer this[long id] => store.LookupCustomer(id);
+
+```
+
+If the method returns `void` or is an async method, the body of the method must be a statement expression (same as with lambdas).  For properties and indexers, they must be read-only, and you do not use the `get` accessor keyword.
+
+## <a name="iterators" /> Iterators ##
+
+An iterator performs a custom iteration over a collection, such as a list or an array. An iterator uses the [yield return](https://msdn.microsoft.com/en-us/library/9k7k7cf0.aspx) statement to return each element one at a time. When a `yield return` statement is reached, the current location in code is remembered. Execution is restarted from that location when the iterator is called the next time.
```

Comment:
The last sentence is misleading. The iterator isn't called a subsequent time, but the caller requests the next element in the sequence (which means a MoveNext is called).